### ISSUE TITLE
Infinite redirection to IDP

### ISSUE THREAD

**Issue by ashutoshkumarmishra560** [dmirandaa/moodle-auth_saml2sso#47]
Hi,

I have configured my Identity Provider with Moodle 3.8.8 using version 3.9 of this plugin and below is the observation:

1. Access Moodle.
2. I get redirected to IDP. I login to IDP
3. SAML token is posted to SimpleSaml and from there it attempts to go to Moodle.
4. Moodle again redirects to IDP.

Below are my SAML auth request and response:
<samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
                    xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
                    ID="_e8f00a76734141ab8152ca793fed9bed6183e3def1"
                    Version="2.0"
                    IssueInstant="2021-04-30T08:40:40Z"
                    Destination="http://localhost:8080/CIDSaas_Sprint32/default/user/spsso"
                    AssertionConsumerServiceURL="http://localhost/simplesaml/module.php/saml/sp/saml2-acs.php/default-sp"
                    ProtocolBinding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
                    >
    <saml:Issuer>http://saml.ilantus.com</saml:Issuer>
    <samlp:NameIDPolicy Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"
                        AllowCreate="true"
                        />
</samlp:AuthnRequest>

<saml2p:Response xmlns:saml2p="urn:oasis:names:tc:SAML:2.0:protocol"
                 Destination="http://localhost:81/simplesaml/module.php/saml/sp/saml2-acs.php/default-sp"
                 ID="IDAAS-3BG7kYSr8NJb12vxC33F"
                 InResponseTo="_e8f00a76734141ab8152ca793fed9bed6183e3def1"
                 IssueInstant="2021-04-30T08:40:48.267Z"
                 Version="2.0"
                 >
    <saml2:Issuer xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">http://issuer.ilantus.com</saml2:Issuer>
    <saml2p:Status>
        <saml2p:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success" />
    </saml2p:Status>
    <saml2:Assertion xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion"
                     xmlns:xs="http://www.w3.org/2001/XMLSchema"
                     ID="IBAM-7RvJlcNtyPCmcu5bxXoj"
                     IssueInstant="2021-04-30T08:40:48.267Z"
                     Version="2.0"
                     >
        <saml2:Issuer Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient">http://issuer.ilantus.com</saml2:Issuer>
        <ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
            <ds:SignedInfo>
                <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" />
                <ds:SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1" />
                <ds:Reference URI="#IBAM-7RvJlcNtyPCmcu5bxXoj">
                    <ds:Transforms>
                        <ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature" />
                        <ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#">
                            <ec:InclusiveNamespaces xmlns:ec="http://www.w3.org/2001/10/xml-exc-c14n#"
                                                    PrefixList="xs"
                                                    />
                        </ds:Transform>
                    </ds:Transforms>
                    <ds:DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1" />
                    <ds:DigestValue>CAzT5T52tg3LlXEoCE4caSTX4SY=</ds:DigestValue>
                </ds:Reference>
            </ds:SignedInfo>
            <ds:SignatureValue>Fq9BocT69bsyCOPUpZM7ZCjpJ9NpSBVhTFFXUSYBXSYF4RPhOc7kA5+1jDarpF44NC/y3Vwx3ryqWrpg1pkwjxKHj75OH1gObkkq7Jfuk61NPaKqW7DVtE8TbXTtLhsWtm2qo2R/AKa0ZBrdxEhEPvCtfLnQKuT/mToILkBNT0c=</ds:SignatureValue>
        </ds:Signature>
        <saml2:Subject>
            <saml2:NameID Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"
                          NameQualifier="http://issuer.ilantus.com"
                          SPNameQualifier="http://saml.ilantus.com"
                          >MPbWshPmlT2ClyEd2ieI</saml2:NameID>
            <saml2:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
                <saml2:SubjectConfirmationData InResponseTo="_e8f00a76734141ab8152ca793fed9bed6183e3def1"
                                               NotOnOrAfter="2021-04-30T08:41:53.267Z"
                                               Recipient="http://localhost:81/simplesaml/module.php/saml/sp/saml2-acs.php/default-sp"
                                               />
            </saml2:SubjectConfirmation>
        </saml2:Subject>
        <saml2:Conditions NotBefore="2021-04-30T08:40:43.267Z"
                          NotOnOrAfter="2021-04-30T08:41:53.267Z"
                          >
            <saml2:AudienceRestriction>
                <saml2:Audience>http://saml.ilantus.com</saml2:Audience>
            </saml2:AudienceRestriction>
        </saml2:Conditions>
        <saml2:AuthnStatement AuthnInstant="2021-04-30T08:40:48.267Z"
                              SessionIndex="v2cGEBwYqsHLUpWrHGkK"
                              SessionNotOnOrAfter="2021-04-30T08:41:53.267Z"
                              >
            <saml2:AuthnContext>
                <saml2:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:Password</saml2:AuthnContextClassRef>
            </saml2:AuthnContext>
        </saml2:AuthnStatement>
        <saml2:AttributeStatement>
            <saml2:Attribute Name="Username">
                <saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                                      xsi:type="xs:string"
                                      >ashutoshkumar.mishra@ilan.com</saml2:AttributeValue>
            </saml2:Attribute>
            <saml2:Attribute Name="firstName">
                <saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                                      xsi:type="xs:string"
                                      >ashutosh</saml2:AttributeValue>
            </saml2:Attribute>
            <saml2:Attribute Name="lastName">
                <saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                                      xsi:type="xs:string"
                                      >admin</saml2:AttributeValue>
            </saml2:Attribute>
            <saml2:Attribute Name="mail">
                <saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                                      xsi:type="xs:string"
                                      >ashutoshkumar.mishra@ilan.com</saml2:AttributeValue>
            </saml2:Attribute>
        </saml2:AttributeStatement>
    </saml2:Assertion>
</saml2p:Response>

Attaching the moodle configuration screenshots below:
![m1](https://user-images.githubusercontent.com/11361005/116674100-dad35f80-a9c1-11eb-9a9a-bef5062b9010.png)
![m2](https://user-images.githubusercontent.com/11361005/116674107-dc048c80-a9c1-11eb-9ac4-1e951ce343cd.png)
![m3](https://user-images.githubusercontent.com/11361005/116674110-dd35b980-a9c1-11eb-8b79-426b0f2be024.png)
![m4](https://user-images.githubusercontent.com/11361005/116674111-dd35b980-a9c1-11eb-8a2d-baf54c9c66e8.png)
![m5](https://user-images.githubusercontent.com/11361005/116674113-ddce5000-a9c1-11eb-8db1-39d03baf45c1.png)

**Comment by mfprimo** [comment]
Hi,
  in the request:

AssertionConsumerServiceURL="http://localhost/simplesaml/module.php/saml/sp/saml2-acs.php/default-sp"

but in the response:

Destination="http://localhost:81/simplesaml/module.php/saml/sp/saml2-acs.php/default-sp"

(with port 81). Even it is possible to handle this situation, are you sure SP cookies pass from the endpoint http://localhost and the 
http://localhost:81 one?

**Comment by ashutoshkumarmishra560** [comment]
Hi,

Thanks for the prompt reply and pointing out the same. So does it mean that SimpleSAML and Moodle can not run in separate instance but have to run in same Apache server?
Pardon my ignorance but I was under the impression that SimpleSAML can exist in its independent instance as Moodle comes with its own independent setup.

**Comment by mfprimo** [comment]
The easiest config is SimpleSAMLphp as a path in the same virtual host of Moodle:

```
<VirtualHost *:443>
        DocumentRoot /your-moodle-root
        ....
        SetEnv SIMPLESAMLPHP_CONFIG_DIR /var/www/simplesamlphp/config
        Alias /simplesaml /var/www/simplesamlphp/www
        <Directory /var/www/simplesamlphp/www>
            Require all granted
        </Directory>
```
with vh-specific ACS/metadata.

You can share an SSP installation and its SP metadata between multiple virtual hosts on the same server providing cookie domains and path are consistens.

**Comment by ashutoshkumarmishra560** [comment]
I now have both SimpleSaml and Moodle running on same server (localhost:81) as pointed out but observation is same. After IDP login the flow comes to SimpleSAML and from there it goes to Moodle and redirects it back to IDP.

**Comment by mfprimo** [comment]
Hi,
  please read closed issues #21, #39, #43, #45 with the same topic: none of them are caused by the plugin, all by a misconfiguration.

Check that:
- do not use phpsession as session store for SimpleSAMLphp: most of PHP applications with a bridge to SSP, including Moodle and Drupal, have an unexpected behviour with this config
- the SSP cookie has path / and a domain shared between the Moodle vh and the SSP vh
- trace SAML messages _and_ HTTP header and send me a dump

**Comment by mfprimo** [comment]
Not a module issue
